### ISSUE TITLE
longStackTraces memory issue

### ISSUE THREAD

**Issue by Volox** [petkaantonov/bluebird#447]
Hi,
i have a memory issue using `Promise.longStackTraces()`. I tried to create an "infite loop" for a crawler and noticed a huge memory usage for the application; i debugged the application and found a strange behaviour.

The following code illustrates the problem:
```js
var Promise = require( 'bluebird' );
Promise.longStackTraces();
var SIZE=1024*1024;

function crawlLoop() {
  var arr = [];
  for( var i=0; i<SIZE; i++ ) {
    arr.push( {} );
  }
  return Promise.resolve( arr ).then( crawlLoop );
}

crawlLoop();
```

After a few tentatives i noticed that by removing `Promise.longStackTraces()` the problem was solved.
Another solution is to call `setImmediate` in the last then of the promise like:
```js
return Promise.resolve( arr ).then( function() { setImmediate( crawlLoop ); } );
```

Can someone please explain me what is wrong with the code (a part for the infinite loop)?

My configuration:
node x64 0.11.13
Windows x64
Bluebird: 2.3.11

Thanks!

**Comment by petkaantonov** [comment]
Can you use the latest version? The memory leak was fixed in [2.4.3](https://github.com/petkaantonov/bluebird/blob/15a75f3d08f11db9562d73f185df417117330b87/changelog.md#243-2014-12-28) (But don't use that exact version, as it caused several regressions which were later fixed)

**Comment by janez87** [comment]
Hi,
I can reproduce the problem with the latest (2.6.4) version.

node x64 0.10.28
osx 10.10

**Comment by Volox** [comment]
Just updated to 2.6.4 still same problem :(

**Comment by petkaantonov** [comment]
I can fix the memory leak but the given code will still keep using maximum memory because it runs synchronously. However if you do `someAsyncAction().then(crawlLoop)` there is no memory issue. Is such fix fine?

**Comment by Volox** [comment]
I've got the same problem also using an async function like in the following code:

```js
var Promise = require( 'bluebird' );
Promise.longStackTraces();

var SIZE=1024*1024;

var asyncFun = function() {
  return new Promise( function( res ) {
    setTimeout( function() {
      var arr = [];
      for( var i=0; i<SIZE; i++ ) {
        arr.push( {} );
      }
      return res( arr );
    }, 100 );
  } );
};

function crawlLoop() {
  return asyncFun()
  .then( crawlLoop );
  // .then( function() { setImmediate( crawlLoop ); } );
}

crawlLoop();
```

**Comment by petkaantonov** [comment]
Fixed in 2.7.1, the original code stays at 100MB working set

**Comment by Volox** [comment]
Thanks!
